artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ActiveMQServerImpl.java

```diff
@@ -1269,6 +1269,8 @@ void stop(boolean failoverOnServerShutdown, final boolean criticalIOError, boole
          this.analyzer = null;
       }
 
+      activateCallbacks.clear();
```

Comment:
do you really need this?

There are a lot of things around activate callbacks. I don't see any testing around this.. so why is this needed?

Reply:
For example, Wildfly will inject activations and a server may stopped and started.

unless you put really good though and testing on this.. you shouldn't really touch this area.

It does not seem related.

Reply:
I will merge without this line.

If you need to clear activateCallbacks, please let me know why and we can review this!

Reply:
Each time the server starts, it register a new activateCallback for FileLockNodeManager and for ManagementServiceImpl. After n stop and start without  `activateCallbacks.clear();` the server has n   activateCallback for FileLockNodeManager  and n  activateCallback for ManagementServiceImpl. After second start setUpServerLockFile is also called on previous activateCallbacks causing file leaks.

Reply:
So the fix should be in the component holding the external resource then. Not here. If a piece of code opens an external resource the logic and lifecycle of that must be managed by it properly. Seems this is just hiding the problem that a component is not managing its rource correctly and simply this is hiding that issur

Reply:
E.g. im worried we are treating symptoms, rather the the underlying cause

Reply:
@michaelandrepearce I see your point, we could replace `activateCallbacks.clear();` with a call to unregisterActivateCallback on LiveOnlyActivation.close for FileLockNodeManager and a call to unregisterActivateCallback on ManagementServiceImpl.stop for ManagementServiceImpl, wdyt?

Reply:
still thats a change here, still not solving the underlying issue is dealing with the component that is holding the FD, it internally should be handling this, NOT ActiveMQServerImpl

Reply:
After start, stop and another start we have two activateCallbacks for FileLockNodeManager one from first start and one from second start. During second server start nodeManager.start/setUpServerLockFile is called on the new FileLockNodeManager but setLive/writeFileLockStatus is also called on previous FileLockNodeManager.
When setLive/writeFileLockStatus is called even on previous FileLockNodeManager the channel is closed (this is right, it is closed on the previous stop) so setUpServerLockFile is called an this is the file leak cause. So I think that who register an activateCallback for FileLockNodeManager on start has to unregister that activateCallback on stop, wdyt?

Reply:
Yes if you add a callback, and its no longer required, the thing/component that does that has the responsibility to remove it, when its not required.

Reply:
Whats the status on this i see this PR was merged is this discussion being addressed separately in a new PR?

Reply:
@michaelandrepearce the callback.clear has been removed while I merged. @brusdev will address the other test that needs a cleanup.

Reply:
I see now. .the issue is that a stop(shutdown=true) should clear the activateCallbacks. 

this is not really an issue when running on a real server, but if you call shutdown=true, it means the VM is going down and the process should cleanup anything left behind.

As the callback is leaking, if you reused the server after a stop(shutdown=true) on the testsuite, you will have the activation (on this case the one holding the lock) being called an extra time, what would then cause the file leak.


I will send another PR with checking the shutdown clause and add a test to avoid a future leak.

Reply:
Query:

Doesnt network checker shutdown and restart in same vm? E.g. would that be liable to leaking?